routing/payment_lifecycle.go

```diff
@@ -26,189 +28,524 @@ func (e errNoRoute) Error() string {
 		e.lastError)
 }
 
+// paymentFailure represents a terminal failure that can happen during the
+// payment lifecycle, indicating that the overall payment should be considered
+// failed. We use it to record such an encountered error while we wait for the
+// remaining shards to finish.
+type paymentFailure struct {
+	failureCode channeldb.FailureReason
+	err         error
+}
+
+// Error returns the human-readable error string for the paymentFailure.
+func (e *paymentFailure) Error() string {
+	return e.err.Error()
+}
+
+// shardOutcome represents the parsed result of a payment shard. It can
+// indicate if the shard succeeded yielding a preimage, if the shard failed
+// with an error that indicates the overall payment should be considered
+// failed, or if only the shard failed and we can re-attempt with the remaining
+// value.
+type shardOutcome struct {
+	// preimage is the preimage retrieved from the succeeding shard. Will
+	// only be non-zero if neither failures are set.
+	preimage lntypes.Preimage
+
+	// paymentFailure indicates this shard encountered a failure serious
+	// enough that we should stop attempting this payment and consider it
+	// failed. If this is nil it is considered safe to make another shard
+	// attempt.
+	paymentFailure *paymentFailure
+
+	// shardFailure is an error indicating that the shard failed. If this
+	// is non-nil but the paymentFailure is nil, we consider the shard
+	// failed, but safe to continue the payment process with new shards.
+	shardFailure error
+}
+
 // paymentLifecycle holds all information about the current state of a payment
 // needed to resume if from any point.
 type paymentLifecycle struct {
-	router         *ChannelRouter
-	payment        *LightningPayment
-	paySession     PaymentSession
-	timeoutChan    <-chan time.Time
-	currentHeight  int32
-	finalCLTVDelta uint16
-	attempt        *channeldb.HTLCAttemptInfo
-	circuit        *sphinx.Circuit
-	lastError      error
+	router        *ChannelRouter
+	totalAmount   lnwire.MilliSatoshi
+	paymentHash   lntypes.Hash
+	paySession    PaymentSession
+	timeoutChan   <-chan time.Time
+	currentHeight int32
 }
 
 // resumePayment resumes the paymentLifecycle from the current state.
 func (p *paymentLifecycle) resumePayment() ([32]byte, *route.Route, error) {
+	shardHandler := &shardHandler{
+		router:      p.router,
+		paymentHash: p.paymentHash,
+		quit:        make(chan struct{}),
+	}
+
+	// When the payment lifecycle loop exits, we make sure to signal any
+	// sub goroutine to exit, then wait for them to return.
+	defer func() {
+		close(shardHandler.quit)
```

Comment:
`resumePayment` quits the shard handler, but below this same function waits for that quit signal. Isn't nothing ever happening in that case?

Reply:
Not entirely sure what you mean?

Reply:
That the same thread that is supposed to close the quit channel also waits on the quit channel. So while waiting on quit, it can never signal quit?

Reply:
Only the sub goroutines wait on the quit channel, not this thread. I can try to extract it more into the `shardHandler` by adding `Start/Stop` methods if that would make it more clear.

Reply:
Ah I see it now. I indeed think it may be better when `shardHandler` manages its own waitgroup.

Reply:
Made `Stop` method to make it more clear.